plugins/scaffolder/src/components/fields/utils.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { JSONSchema7 } from 'json-schema';
+import { z } from 'zod';
+import zodToJsonSchema from 'zod-to-json-schema';
+
+/**
+ * @public
+ * Utility function to convert zod schemas to JSON schemas with
+ * type inference extraction that abstracts away zod typings
+ */
+export function makeJsonSchemaFromZod<T extends z.ZodType>(
+  schema: T,
+): {
+  schema: JSONSchema7;
+  type: T extends z.ZodType<any, any, infer I> ? I : never;
+} {
+  return {
+    schema: zodToJsonSchema(schema) as JSONSchema7,
+    type: null as any,
+  };
+}
```

Comment:
```suggestion
/**
 * @public
 * FieldSchema encapsulates a JSONSchema7 along with the
 * matching input type for a field extension.
 */
export interface FieldSchema<TInput> {
  readonly schema: JSONSchema7;
  readonly type: TInput;
}

/**
 * @public
 * Utility function to convert zod schemas to JSON schemas with
 * type inference extraction that abstracts away zod typings
 */
export function makeFieldSchemaFromZod<T extends z.ZodType>(
  schema: T,
): FieldSchema<T extends z.ZodType<any, any, infer I> ? I : never> {
  return {
    schema: zodToJsonSchema(schema) as JSONSchema7,
    type: null as any,
  };
}
```

Was thinking we could encapsulate this a bit perhaps to clean things up even further? Ends up pretty nice in the API report too.

Looking at this I realized there might be a tiny discrepancy too btw:

```ts
export const RepoUrlPickerSchema = {
  uiOptions: RepoUrlPickerUiOptionsSchema.schema,
  returnValue: RepoUrlPickerReturnValueSchema.schema,
};
```

The current util function grabs the input type, but I'm assuming that for the return value we actually want to be grabbing the output type instead?

Taking this to an extreme, do you think we could take this all the way to a utility that creates the `FieldExtensionOptions` type for the two schemas? I'm still feeling there are quite a for intermediary types and helpers that could perhaps be avoided.

Reply:
Good idea! Cleaned up the change by expanding the helper function. Thanks for the suggestion!